File: client-sdk-ts/src/loader.ts
/*
 * Copyright 2025 promptLM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { PromptArtifactError, PromptNotFoundError } from "./errors.js";
import type { Prompt, PromptIndexEntry, PromptLoader, PromptSource } from "./types.js";

const INDEX_PATH = "prompts/prompt-index.json";

function requireString(value: unknown, field: string, source: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new PromptArtifactError(`Missing or invalid ${field} in ${source}`);
  }
  return value;
}

export class JsonPromptLoader implements PromptLoader {
  private indexPromise?: Promise<Map<string, PromptIndexEntry>>;

  constructor(private readonly source: PromptSource) {}

  async loadPrompt(id: string): Promise<Prompt> {
    const index = await this.loadIndex();
    const entry = index.get(id);
    if (!entry) {
      throw new PromptNotFoundError(id);
    }

    const rawPrompt = await this.source.readText(entry.path);
    let parsed: unknown;
    try {
      parsed = JSON.parse(rawPrompt);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new PromptArtifactError(`Invalid JSON in ${entry.path}: ${message}`);
    }

    if (!parsed || typeof parsed !== "object") {
      throw new PromptArtifactError(`Prompt file ${entry.path} must contain a JSON object`);
    }

    const promptRecord = parsed as Record<string, unknown>;
    return {
      id: requireString(promptRecord.id, "id", entry.path),
      version: requireString(promptRecord.version, "version", entry.path),
      name: requireString(promptRecord.name, "name", entry.path),
      path: entry.path,
      prompt: requireString(promptRecord.prompt, "prompt", entry.path)
    };
  }

  private loadIndex(): Promise<Map<string, PromptIndexEntry>> {
    if (!this.indexPromise) {
      this.indexPromise = this.readIndex();
    }
    return this.indexPromise;
  }

  private async readIndex(): Promise<Map<string, PromptIndexEntry>> {
    const rawIndex = await this.source.readText(INDEX_PATH);
    let parsed: unknown;
    try {
      parsed = JSON.parse(rawIndex);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new PromptArtifactError(`Invalid JSON in ${INDEX_PATH}: ${message}`);
    }

    if (!Array.isArray(parsed)) {
      throw new PromptArtifactError(`${INDEX_PATH} must contain a JSON array`);
    }

    const index = new Map<string, PromptIndexEntry>();
    for (const candidate of parsed) {
      if (!candidate || typeof candidate !== "object") {
        throw new PromptArtifactError(`${INDEX_PATH} contains an invalid entry`);
      }

      const entryRecord = candidate as Record<string, unknown>;
      const entry: PromptIndexEntry = {
        id: requireString(entryRecord.id, "id", INDEX_PATH),
        version: requireString(entryRecord.version, "version", INDEX_PATH),
        name: requireString(entryRecord.name, "name", INDEX_PATH),
        path: requireString(entryRecord.path, "path", INDEX_PATH)
      };

      if (index.has(entry.id)) {
        throw new PromptArtifactError(`Duplicate prompt id in ${INDEX_PATH}: ${entry.id}`);
      }

      index.set(entry.id, entry);
    }

    return index;
  }
}

File: client-sdk-python/src/promptlm_client/loader.py
# Copyright 2025 promptLM
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import json
from typing import Any

from .errors import PromptArtifactError, PromptNotFoundError
from .models import Prompt

INDEX_PATH = "prompts/prompt-index.json"


def _require_string(value: Any, field_name: str, source: str) -> str:
    if not isinstance(value, str) or not value:
        raise PromptArtifactError(f"Missing or invalid {field_name} in {source}")
    return value


class JsonPromptLoader:
    def __init__(self, source: Any) -> None:
        self._source = source
        self._index: dict[str, dict[str, str]] | None = None

    def load_prompt(self, prompt_id: str) -> Prompt:
        index = self._load_index()
        entry = index.get(prompt_id)
        if entry is None:
            raise PromptNotFoundError(prompt_id)

        raw_prompt = self._source.read_text(entry["path"])
        try:
            parsed = json.loads(raw_prompt)
        except json.JSONDecodeError as error:
            raise PromptArtifactError(f"Invalid JSON in {entry['path']}: {error}") from error

        if not isinstance(parsed, dict):
            raise PromptArtifactError(f"Prompt file {entry['path']} must contain a JSON object")

        return Prompt(
            id=_require_string(parsed.get("id"), "id", entry["path"]),
            version=_require_string(parsed.get("version"), "version", entry["path"]),
            name=_require_string(parsed.get("name"), "name", entry["path"]),
            path=entry["path"],
            prompt=_require_string(parsed.get("prompt"), "prompt", entry["path"]),
        )

    def _load_index(self) -> dict[str, dict[str, str]]:
        if self._index is None:
            self._index = self._read_index()
        return self._index

    def _read_index(self) -> dict[str, dict[str, str]]:
        raw_index = self._source.read_text(INDEX_PATH)
        try:
            parsed = json.loads(raw_index)
        except json.JSONDecodeError as error:
            raise PromptArtifactError(f"Invalid JSON in {INDEX_PATH}: {error}") from error

        if not isinstance(parsed, list):
            raise PromptArtifactError(f"{INDEX_PATH} must contain a JSON array")

        index: dict[str, dict[str, str]] = {}
        for entry in parsed:
            if not isinstance(entry, dict):
                raise PromptArtifactError(f"{INDEX_PATH} contains an invalid entry")

            candidate = {
                "id": _require_string(entry.get("id"), "id", INDEX_PATH),
                "version": _require_string(entry.get("version"), "version", INDEX_PATH),
                "name": _require_string(entry.get("name"), "name", INDEX_PATH),
                "path": _require_string(entry.get("path"), "path", INDEX_PATH),
            }

            if candidate["id"] in index:
                raise PromptArtifactError(
                    f"Duplicate prompt id in {INDEX_PATH}: {candidate['id']}"
                )

            index[candidate["id"]] = candidate

        return index
